### PR TITLE
Check that duplicate Orchard nullifiers within transactions are rejected

### PR DESCRIPTION
**Pull request by ebfull**

### CODE REVIEW

File: src/gtest/test_transaction_builder.cpp [hunk 1]
```diff
@@ -201,6 +201,27 @@ TEST(TransactionBuilder, SproutToSproutAndSapling) {
...
+
+    CTransaction tx;
+    ssin >> tx;
+
+    CValidationState state;
+    EXPECT_FALSE(CheckTransactionWithoutProofVerification(tx, state));
+    EXPECT_EQ(state.GetRejectReason(), "bad-orchard-nullifiers-duplicate");
+
+    // Revert to default
+    RegtestDeactivateNU5();
```
Comment (nuttycom) [inline src/gtest/test_transaction_builder.cpp:222]: Why does this deactivate without a corresponding activation?

**Review by nuttycom — CHANGES_REQUESTED** [review]

**Comment by ebfull** [comment]
FYI here are the changes to the `orchard` crate I made for this PR locally.

```
diff --git a/src/builder.rs b/src/builder.rs
index 4a207c2..721d39a 100644
--- a/src/builder.rs
+++ b/src/builder.rs
@@ -51,7 +51,7 @@ impl From<value::OverflowError> for Error {
 }
 
 /// Information about a specific note to be spent in an [`Action`].
-#[derive(Debug)]
+#[derive(Debug, Clone)]
 struct SpendInfo {
     dummy_sk: Option<SpendingKey>,
     fvk: FullViewingKey,
@@ -224,6 +224,17 @@ impl Builder {
         }
     }
 
+    /// lol
+    pub fn add_duplicate_nullifier_spend(
+        &mut self,
+        mut rng: impl RngCore,
+    )
+    {
+        let spend = SpendInfo::dummy(&mut rng);
+        self.spends.push(spend.clone());
+        self.spends.push(spend);
+    }
+
     /// Adds a note to be spent in this transaction.
     ///
     /// Returns an error if the given Merkle path does not have the required anchor for
diff --git a/src/tree.rs b/src/tree.rs
index 874d08e..305e44f 100644
--- a/src/tree.rs
+++ b/src/tree.rs
@@ -75,7 +75,7 @@ impl Anchor {
 
 /// The Merkle path from a leaf of the note commitment tree
 /// to its anchor.
-#[derive(Debug)]
+#[derive(Debug, Clone)]
 pub struct MerklePath {
     position: u32,
     auth_path: [MerkleHashOrchard; MERKLE_DEPTH_ORCHARD],
```

**Review by nuttycom — APPROVED** [review]
utACK

**Review by therealyingtong — APPROVED** [review]
utACK

**Comment by nuttycom** [comment]
CI failures are unrelated nondeterministic RPC test issues.

File: src/gtest/data/tx-orchard-duplicate-nullifiers.h [hunk 2]
```diff
@@ -0,0 +1,4 @@
...
...55ef26839dea0fd409e9b92602b1f11a747a5f7d6b88089c6a58c1b0ec298554219084475f05145aedf0284c4b97997911660ee8dd70b1c7d3e8bd15b9022ef54066c9c31362338d1a92092fe00a80f09348cb488416305997403cb4df14a7f9ae156f96717e6b661be83c7df33f14a1f2ec41b8c774a0b0cdaa222efcfaf838f7ba0d7db202a8fc227e869f631fd04afb1ec06aa5511b117af5678edb698d0e4eb7c7f0cc3fe52644a6f8d58f720c0621627fce7099af0789d50cf8e7aea67cd395c17c234ce5a62133525ba528f07f1bed690fdf53af3f806140255f4dd9a5190b469809e67e762d227587f48f29e6d0915d2f5852b0d56001abd5058bd84f4ca989474820a46a515204b81b6fabc39d3f0c3ca196892e1cae1d068f0555f301d84af9c43e9545316d951b132e931bf561ea12f01a20e7b855a385fb6f8bd5e8640702c261c17707864f211ff81ca3e1f3782b04d69e33ee93ec286a1873b82ad20a43e1d79d8fd3fa9820383ba2bf45133d5a621e3145c3c70c0f5e9faddfe65d109735cf60096b868a2ae86feb61a1d2f92a190108b7d31ded1cd3ff4d726c6f3dd68a3fbe50b720273efba99d85b987a2cf1567957f373cd4251f98b632643481c7d089fb3d150f0505355d6412e4d2ddd1ab611da3b03819c524ba043f1b155ef36033497583b4804cd18ce0ddb855ce027e5fa1e3d7e5bdd2793bc5b20cf417887e40761c0e9f46a305fd5098bf5f5d8fcb303ebc52c091c9201f945d2627f1184c3aae5e8420f874b52844d35e6392a767e19f2996f18050bcea6fd413adf42281bad5229f0544bb3bb339e3e894f8f21ee22f";
```
Comment (str4d) [inline src/gtest/data/tx-orchard-duplicate-nullifiers.h:3]: I verified that this does indeed contain two Orchard actions that use the same nullifier (and a valid proof, indicating the actions double-spend the same note).

**Review by str4d — APPROVED** [review]
utACK. I didn't check the test in this PR passes, but I did update my `zcash-inspect` tool to check this consensus rule.

**Review by daira — COMMENTED** [review]
Post-hoc ACK

**Pull request merged**